src/osd/modules/debugger/win/consolewininfo.cpp

```diff
@@ -20,10 +20,185 @@
 #include "strconv.h"
 #include "winutf8.h"
 
+#include <string>
 
-consolewin_info::consolewin_info(debugger_windows_interface &debugger) :
-	disasmbasewin_info(debugger, true, "Debug", nullptr),
-	m_devices_menu(nullptr)
+using namespace std::literals;
+
+
+//**************************************************************************
+//  ANONYMOUS NAMESPACE
+//**************************************************************************
+
+namespace
+{
+
+//-------------------------------------------------
+//  copy_extension_list
+//-------------------------------------------------
+
+void copy_extension_list(std::ostringstream &dest, const char *extensions)
+{
+	// our extension lists are comma delimited; Win32 expects to see lists
+	// delimited by semicolons
+	char const *s = extensions;
+	while (*s)
+	{
+		// append a semicolon if not at the beginning
+		if (s != extensions)
+			dest << ';';
+
+		// append ".*"
+		dest << "*.";
+
+		// append the file extension
+		while (*s && (*s != ','))
+			dest << *s++;
+
+		// if we found a comma, advance
+		while (*s == ',')
+			s++;
+	}
+}
+
+
+//-------------------------------------------------
+//  add_filter_entry
+//-------------------------------------------------
+
+void add_filter_entry(std::ostringstream &dest, const char *description, const char *extensions)
+{
+	// add the description
+	dest << description;
+	dest << " (";
+
+	// add the extensions to the description
+	copy_extension_list(dest, extensions);
+
+	// add the trailing rparen and '\0' character
+	dest << ")\0"s;
+
+	// now add the extension list itself
+	copy_extension_list(dest, extensions);
+
+	// append a '\0'
+	dest << '\0';
+}
+
+
+//-------------------------------------------------
+//  build_generic_filter
+//-------------------------------------------------
+
+std::string build_generic_filter(device_image_interface *img, bool is_save)
+{
+	std::ostringstream filter;
+	std::string file_extension;
+
+	if (img)
+		file_extension = img->file_extensions();
+
+	if (!is_save)
+		file_extension.append(",zip,7z");
+
+	add_filter_entry(filter, "Common image types", file_extension.c_str());
+
+	filter << "All files (*.*)\0*.*\0"s;
+
+	if (!is_save)
+		filter << "Compressed Images (*.zip;*.7z)\0*.zip;*.7z\0"s;
+
+	return filter.str();
+}
+
+
+//-------------------------------------------------
+//  win_get_file_name
+//-------------------------------------------------
+
+enum class win_get_file_name_type
+{
+	OPEN,
+	SAVE
+};
+
+std::string win_get_file_name(win_get_file_name_type type, const std::string &filter, const std::string &dir)
+{
+	// convert to TCHAR
+	osd::text::tstring t_filter = osd::text::to_tstring(filter);
+	osd::text::tstring t_dir = osd::text::to_tstring(dir);
+
+	// this is where the selected filename is stored
+	TCHAR selected_filename[MAX_PATH];
+	selected_filename[0] = '\0';
+
+	// record the current working directory
+	DWORD working_directory_length = GetCurrentDirectory(0, nullptr);
```

Comment:
This isn't thread-safe since the working directory is per-process.  The web server and lua engine are on work threads, so they'll still get screwed up by this.